rust/otap-dataflow/crates/otap/src/pdata.rs

```diff
@@ -0,0 +1,310 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+use otap_df_pdata_views::otlp::bytes::logs::RawLogsData;
+use otel_arrow_rust::otap::{OtapBatch, from_record_messages};
+use otel_arrow_rust::otlp::{logs::logs_from, metrics::metrics_from, traces::traces_from};
+use otel_arrow_rust::{Consumer, Producer};
+use prost::{EncodeError, Message};
+
+use crate::{encoder::encode_logs_otap_batch, grpc::OTAPData};
+
+/// module contains related to pdata
+pub mod error {
+    /// Errors related to pdata
+    #[derive(thiserror::Error, Debug)]
+    pub enum Error {
+        /// Wrapper for error that occurred converting pdata
+        #[error("An error occurred converting pdata: {error}")]
+        ConversionError {
+            /// The error that occurred
+            error: String,
+        },
+    }
+
+    impl<T> From<Error> for otap_df_engine::error::Error<T> {
+        fn from(e: Error) -> Self {
+            otap_df_engine::error::Error::PdataConversionError {
+                error: format!("{e}"),
+            }
+        }
+    }
+}
+
+/// Pipeline data represented as protobuf serialized OTLP request messages
+#[derive(Clone, Debug)]
+pub enum OtlpProtoBytes {
+    /// protobuf serialized ExportLogsServiceRequest
+    ExportLogsRequest(Vec<u8>),
+    /// protobuf serialized ExportMetricsServiceRequest
+    ExportMetricsRequest(Vec<u8>),
+    /// protobuf serialized ExportTracesServiceRequest
+    ExportTracesRequest(Vec<u8>),
+}
+
+/// Container for the various representations of the telemetry data
+#[derive(Clone, Debug)]
+#[allow(clippy::large_enum_variant)]
+pub enum OtapPdata {
+    /// data is serialized as a protobuf service message for one of the OTLP GRPC services
+    OtlpBytes(OtlpProtoBytes),
+
+    /// data is contained in `BatchArrowRecords`, which contain ArrowIPC serialized
+    OTAPData(OTAPData),
+
+    /// data is contained in `OtapBatch` which contains Arrow `RecordBatches` for OTAP payload type
+    OtapBatch(OtapBatch),
+}
```

Comment:
I wonder if we shouldn’t make the following renamings to clarify things:
- `OTAPData` -> `OtapArrowBytes`
- `OtapBatch` -> `OtapArrowRecords`

Reply:
sounds good, made this change